mne/io/nsx/nsx.py

```diff
@@ -0,0 +1,601 @@
+
+@fill_doc
+class RawNSX(BaseRaw):
+    """Raw object from NSx file from Blackrock Microsystems.
+
+    Parameters
+    ----------
+    input_fname : str
+        Path to the NSx file.
+    stim_channel : ``'auto'`` | str | list of str | int | list of int
+        Defaults to ``'auto'``, which means that channels named ``'status'`` or
+        ``'trigger'`` (case insensitive) are set to STIM. If str (or list of
+        str), all channels matching the name(s) are set to STIM. If int (or
+        list of ints), channels corresponding to the indices are set to STIM.
+    eog : list or tuple
+        Names of channels or list of indices that should be designated EOG
+        channels. Values should correspond to the electrodes in the file.
+        Default is None.
+    misc : list or tuple
+        Names of channels or list of indices that should be designated MISC
+        channels. Values should correspond to the electrodes in the file.
+        Default is None.
+    %(preload)s
+    %(verbose)s
+
+    Notes
+    -----
+    NSx files with id (= NEURALSG), i.e., version 2.1 is currently not
+    supported.
+
+    If channels named 'status' or 'trigger' are present, they are considered as
+    STIM channels by default. Use func:`mne.find_events` to parse events
+    encoded in such analog stim channels.
+    """
+
+    def __init__(
+        self,
+        input_fname,
+        stim_channel="auto",
+        eog=None,
+        misc=None,
+        preload=False,
+        verbose=None,
+    ):
+        logger.info("Extracting NSX parameters from {}...".format(input_fname))
+        input_fname = os.path.abspath(input_fname)
+        (
+            info,
+            data_fname,
+            fmt,
+            n_samples,
+            orig_format,
+            raw_extras,
+            orig_units,
+        ) = _get_hdr_info(input_fname, stim_channel=stim_channel, eog=eog, misc=misc)
+        raw_extras["orig_format"] = orig_format
+        super(RawNSX, self).__init__(
+            info,
+            last_samps=[n_samples - 1],
+            filenames=[data_fname],
+            orig_format=orig_format,
+            preload=preload,
+            verbose=verbose,
+            raw_extras=[raw_extras],
+            orig_units=orig_units,
+        )
+
+        # Add annotations for in-data skips
+        if len(self._raw_extras[0]["timestamp"]) > 1:
+            starts = (
+                self._raw_extras[0]["timestamp"] + self._raw_extras[0]["nb_data_points"]
+            )[:-1] + 1
+            stops = self._raw_extras[0]["timestamp"][1:] - 1
+            durations = (stops - starts + 1) / self.info["sfreq"]
+            annot = Annotations(
+                onset=(starts / self.info["sfreq"]),
+                duration=durations,
+                description="BAD_ACQ_SKIP",
+                orig_time=self.info["meas_date"],
+            )
+            self.set_annotations(annot)
+
+    def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
```

Comment:
:eyes: There is a lot of technical stuff going on here with skips, header data types, etc. It all looks well coded (nice!) but I'm going to have to trust that you've thought it through and that the tests + use cases you have so far have helped validate it. But let me know if there are specific parts you weren't/aren't sure about that could use additional feedback

Reply:
I think I got this. There might be more efficient way of doing this, but so far this seem to work with my curated data.

Reply:
Okay now that you use our built-in `_read_segments_file` it's much more readable!